N5 root array fix

d-v-b opened the pull request:
Adds a routine to `__setitem__` for `N5Store` and `N5FSStore` that inserts the `n5` keyword to arrays that happen to exist at the root of the hierarchy, and adds tests for this behavior.

Fixes #1092 


TODO:
* [x] Add unit tests and/or doctests in docstrings
* [x] Add docstrings and API docs for any new/modified user-facing classes and functions
* [] New/modified features documented in docs/tutorial.rst
* [x] Changes documented in docs/release.rst
* [x] GitHub Actions have all passed
* [x] Test coverage is 100% (Codecov passes)

pep8speaks commented:
Hello @d-v-b! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



* In the file [`zarr/tests/test_storage.py`](https://github.com/zarr-developers/zarr-python/blob/1b2d1404cda090c73ccc98b0cb8a3dd41ba43913/zarr/tests/test_storage.py):

> [Line 1512:49](https://github.com/zarr-developers/zarr-python/blob/1b2d1404cda090c73ccc98b0cb8a3dd41ba43913/zarr/tests/test_storage.py#L1512): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 1636:49](https://github.com/zarr-developers/zarr-python/blob/1b2d1404cda090c73ccc98b0cb8a3dd41ba43913/zarr/tests/test_storage.py#L1636): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace

d-v-b commented:
cc @GenevieveBuckley

d-v-b commented:
I went kind of crazy and removed a lot of the old `# pragma: no cover` statements and added corresponding tests. Locally I'm at 100% coverage for `n5.py`.

d-v-b commented:
the automated tests are timing out, @jakirkham any idea why that might be?

joshmoore commented:
@d-v-b, _lots_ of goodness in this. Would you suggest this appropriate for an immediate 2.12.1?

d-v-b commented:
@joshmoore i need to double-check that I didn't break something here: https://github.com/zarr-developers/zarr-python/pull/1094/files#diff-427748c623f9d38d332c4b0e7e733c52bad927fee7b7ad80c36789aa85a8e142R928

This code block was excluded from testing before, and I changed it to pass tests (specifically, I call `.tobytes()` on the output instead of returning a numpy array). This makes the routine type-stable, but I need to make sure that it doesn't break stuff on n5 datasets generated by the java n5 library. I will let you know when that testing is finished.

jakirkham commented on zarr/n5.py:
```diff
@@ -889,13 +919,13 @@ def decode(self, chunk, out=None):
             chunk = self._from_big_endian(chunk)
 
             # read partial chunk
-            if chunk_shape != self.chunk_shape:  # pragma: no cover
+            if chunk_shape != self.chunk_shape:
                 chunk = np.frombuffer(chunk, dtype=self.dtype)
                 chunk = chunk.reshape(chunk_shape)
                 complete_chunk = np.zeros(self.chunk_shape, dtype=self.dtype)
                 target_slices = tuple(slice(0, s) for s in chunk_shape)
                 complete_chunk[target_slices] = chunk
-                chunk = complete_chunk
+                chunk = complete_chunk.tobytes()
```
Could you please expand a bit on why this is needed? Mentioning as this effectively doing another allocation/copy here

d-v-b replied:
I don't think it's needed, and I removed it

jakirkham reviewed (commented):
Thanks Davis! 🙏

Had a question below

d-v-b commented:
@joshmoore I reverted the change I was worried about, so I think this is ready to go

GenevieveBuckley commented:
Thank you for working on this! However, I don't think this quite matches the N5 spec. 

I've just tried making an N5 file with this PR branch, and the output cannot be read in by the Paintera software. I'm no expert on the N5 spec, but Paintera was developed by the same group that developed the N5 format, so I feel like this is a good test. The problem is that there is not a separate dataset subdirectory, instead the dataset is left loose in the top level root directory. 

Caleb [says in a comment here](https://github.com/saalfeldlab/paintera/issues/457#issuecomment-1155816174) that generally it is expected that N5 directories have this format:
```
container.n5
│   attributes.json         <--- this contains the n5 version tag
└───data
     │   attributes.json    <--- this is the file you posted above
     │   .... (actual data) 
```

What we get from this PR is instead:
```
container.n5
│   attributes.json    <--- now with an extra n5 version tag in the dictionary
│   .... (actual data) 
```

GenevieveBuckley commented:
Wait, no I **can** open it in Paintera! It's just that:
1. I needed to create a 3D dummy dataset, and 
2. It is extremely difficult to position the mouse to click on the single character `/` that is the root dataset. Mostly it just looks like the dataset dropdown menu is blank & if you do try to click it won't select the root dataset unless your mouse is in *just* the right position. So it's easy to assume there is nothing there, or that the file cannot be opened.

GenevieveBuckley commented:
Both of these examples work now, yay 🎉 

```python
In [1]: import numpy as np
In [2]: import zarr

In [3]: data_shape = (1000, 1000, 1000)
In [4]: data_chunks = (100, 100, 100)
In [5]: dummy_data = np.random.randint(0, 255, data_shape)

# Example 1: Group dataset
In [6]: store = zarr.N5Store('group.n5')
In [7]: root = zarr.group(store=store, overwrite=True)
In [8]: foo = root.create_group('foo')
In [9]: bar = foo.zeros('bar', shape=data_shape, chunks=data_chunks)
In [10]: bar[:] = dummy_data

# Example 2: a single dataset stored in the root
In [11]: store = zarr.N5Store('array.n5')
In [12]: z = zarr.zeros((10,10, 10), chunks=(5, 5, 5), store=store, overwrite=True)
In [13]: z[...] = 42
```

I am able to open both `array.n5` and `group.n5` with Paintera. Thank you @d-v-b

GenevieveBuckley commented:
One thing I did not actually check is whether this works just as well with `N5FSStore` as it does with `N5Store`. I'm guessing it will, since it's the `N5ChunkWrapper` that has these changes. Just letting you know I didn't explicitly test that out myself.

joshmoore commented:
Thanks, @GenevieveBuckley. Merging. My plan (as I mentioned on [gitter](https://gitter.im/zarr-developers/community)) was to target this for a 2.13 pre-release. If anyone needs a 2.12 patch release, please let me know.

joshmoore commented on docs/release.rst:
```diff
@@ -29,6 +29,13 @@ Maintenance
+* Fix bug in N5 storage that prevented arrays located in the root of the hierarchy from
+  bearing the `n5` keyword. Along with fixing this bug, new tests were added for N5 routines
+  that had previously been excluded from testing, and type annotations were added to the N5 codebase.
+  By :user:`Davis Bennett <d-v-b>`; :issue:`1092`.
```
```suggestion
  By :user:`Davis Bennett <d-v-b>`; :issue:`1092`.

```
